::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
: Copyright (c) 2018 Mirantis Inc., Enea AB and others.
:
: All rights reserved. This program and the accompanying materials
: are made available under the terms of the Apache License, Version 2.0
: which accompanies this distribution, and is available at
: http://www.apache.org/licenses/LICENSE-2.0
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
From: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
Date: Wed, 22 Aug 2018 01:35:06 +0200
Subject: [PATCH] OPNFV package installation, Ubuntu user

* Install OpenSSH server (and client), so other OPNFV projects can
  easily connect to the Salt master node;
* While at it, create 'ubuntu' user so other OPNFV projects don't
  have to switch to 'root' login;
* Preinstall `salt_minion_dependency_packages` and
  `salt_minion_reclass_dependencies`;
* Fetch opnfv.tar if OPNFV_EXTRA URI is set (to be used for injecting
  OPFNV formulas and reclass.system git repos during build);
* Use formula-fetch.sh from pending PR #38 [1] that allows local FS
  and/or Gerrit URIs for salt formula sources;

[1] https://github.com/salt-formulas/salt-formulas-scripts/pull/38

Signed-off-by: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
---
 DockerMake.yml | 36 ++++++++++++++++++++++++++++++++++--
 Pipfile        |  1 +
 2 files changed, 35 insertions(+), 2 deletions(-)

diff --git a/DockerMake.yml b/DockerMake.yml
index 2c75586..48f1d23 100644
--- a/DockerMake.yml
+++ b/DockerMake.yml
@@ -29,6 +29,7 @@ common-cleanup:

 common:
   build: |
+    ARG CACHE_INVALIDATE="0"
     RUN echo "Layer with common packages" \
       && eval ${LAYER_PKGUPDT} \
       && apt-get upgrade -qy \
@@ -102,12 +103,37 @@ salt-formulas:
     ENV SALT_ENV_PATH_ $SALT_ENV_PATH_
     ARG RECLASS_BASE="/srv/salt/reclass"
     ENV RECLASS_BASE $RECLASS_BASE
+    ARG OPNFV_EXTRA=""
+    ENV OPNFV_EXTRA $OPNFV_EXTRA
     RUN echo "Layer python/salt module prerequisites, formulas" \
+      && test -z "$OPNFV_EXTRA" || curl -sSqL $OPNFV_EXTRA | tar xf - -C /tmp \
       && mkdir -p /srv/salt \
-      && curl -sSqL https://raw.githubusercontent.com/salt-formulas/salt-formulas-scripts/master/formula-fetch.sh -o /srv/salt/formula-fetch.sh \
+      && curl -sSqL https://raw.githubusercontent.com/salt-formulas/salt-formulas-scripts/252ddc6/formula-fetch.sh -o /srv/salt/formula-fetch.sh \
       && bash -c 'source /srv/salt/formula-fetch.sh && setupPyEnv && fetchAll' \
       && eval ${LAYER_CLEANUP}

+opnfv:
+  build: |
+    RUN echo "Layer with OPNFV packages" \
+      && eval ${LAYER_PKGUPDT} \
+      && ${LAYER_INSTALL} \
+           gawk \
+           inetutils-ping \
+           kmod \
+           net-tools \
+           openssh-server \
+           python-futures \
+           python-m2crypto \
+           python-msgpack \
+           python-netaddr \
+           python-oauth \
+           python-psutil \
+           python-yaml \
+      && useradd -m ubuntu \
+      && echo 'ubuntu ALL=(ALL) NOPASSWD:ALL' > /etc/sudoers.d/ubuntu \
+      && eval ${LAYER_CLEANUP}
+
+
 wheel:
   requires:
     - base
@@ -138,9 +164,15 @@ reclass:
     ENV SALT_ENV_PATH_ $SALT_ENV_PATH_
     ARG RECLASS_BASE="/srv/salt/reclass"
     ENV RECLASS_BASE $RECLASS_BASE
+    ARG RECLASS_BASE_SOURCE="https://github.com/Mirantis/reclass-system-salt-model"
+    ENV RECLASS_BASE_SOURCE $RECLASS_BASE_SOURCE
+    ARG OPNFV_EXTRA=""
+    ENV OPNFV_EXTRA $OPNFV_EXTRA
     RUN echo "Layer reclass" \
+      && test -z "$OPNFV_EXTRA" || curl -sSqL $OPNFV_EXTRA | tar xf - -C /tmp \
       && mkdir -p /etc/reclass $RECLASS_BASE/classes/system \
-      && git clone https://github.com/Mirantis/reclass-system-salt-model $RECLASS_BASE/classes/system \
+      && git clone --no-hardlinks $RECLASS_BASE_SOURCE $RECLASS_BASE/classes/system \
+      && rm -rf /tmp/opnfv \
       && find $RECLASS_BASE/classes/system
     RUN pip install --install-option="--prefix=/usr/local" -I \
            "git+https://github.com/salt-formulas/reclass.git@$RECLASS_VERSION"
diff --git a/Pipfile b/Pipfile
index d3e8d66..6067dd2 100644
--- a/Pipfile
+++ b/Pipfile
@@ -9,6 +9,7 @@ name = "pypi"
 pygithub = "*"
 docker-make = {git = "https://github.com/avirshup/DockerMake"}
 dockermake = {git = "https://github.com/avirshup/DockerMake"}
+invoke = "*"

 [requires]
 python_version = "3.6"
